grive2: init at 0.5.0

guaraqe opened the pull request:
The package `grive` has been deprecated and does not work anymore, this is its successor.

Package built with NixOS 15.09. I tried sandboxing but I'm not sure if I managed, since it was already built once.

The binaries have been tested and work as expected.

Commit:
grive2: init at 0.5.0

grive2: update 0.5.0

grive2: various cosmetic changes

joachifm commented on pkgs/tools/filesystems/grive2/default.nix:
```diff
@@ -0,0 +1,26 @@
+
+  buildInputs = [cmake pkgconfig libgcrypt yajl curl expat stdenv binutils boost];
+
+  buildPhase = ''
```
Is it really necessary to specify a build phase?

joachifm commented on pkgs/tools/filesystems/grive2/default.nix:
```diff
@@ -0,0 +1,26 @@
+
+  src = fetchurl {
+    url = https://github.com/vitalif/grive2/archive/v0.5.0.tar.gz;
+    md5 = "f291ffaef831abf3605417995082bcc3";
```
Please use sha256

joachifm commented on pkgs/tools/filesystems/grive2/default.nix:
```diff
@@ -0,0 +1,26 @@
+  version = "0.5.0";
+  name = "grive2-${version}";
+
+  src = fetchurl {
```
Please use `fetchFromGitHub` instead

joachifm commented on pkgs/tools/filesystems/grive2/default.nix:
```diff
@@ -0,0 +1,26 @@
+    md5 = "f291ffaef831abf3605417995082bcc3";
+  };
+
+  buildInputs = [cmake pkgconfig libgcrypt yajl curl expat stdenv binutils boost];
```
`stdenv` and `binutils` are redundant `buildInputs`; also, strictly build-time dependencies such as `cmake` and `pkgconfig` can be moved to `nativeBuildInputs`.

joachifm commented on pkgs/tools/filesystems/grive2/default.nix:
```diff
@@ -0,0 +1,26 @@
+
+  buildPhase = ''
+  cmake .
+  make -j4
```
Nix already takes care of parallel builds, passing `-j` does not make sense

joachifm commented on pkgs/tools/filesystems/grive2/default.nix:
```diff
@@ -0,0 +1,26 @@
+    description = "An open source Linux client for Google Drive";
+    homepage =https://github.com/vitalif/grive2;
+    license = stdenv.lib.licenses.gpl2;
+    platforms = stdenv.lib.platforms.all;
```
According to the description, this is linux only, so maybe use `platforms.linux` here.

joachifm commented:
To force a re-build (e.g., with sandboxing enabled), you can do `nix-store --delete result` where result is a store path.

guaraqe commented:
So I:
- changed to `fetchFromGitHub` and `sha256`
- removed `buildPhase`
- changed the platform to `linux`
- built with sandboxing

I tried to remove either `stdenv` or `binutils` from `buildInputs`, but without binutils it didn't build and without `stdenv` it had runtime problems (I don't know how to explain).

joachifm commented:
Would you please provide a gist/paste of the build error and the run-time error?

guaraqe commented:
The build error (no `binutils`) is due to the lack of `bfd.h`.

The runtime error seems to be related to `boost`:
```
exception: /tmp/nix-build-grive2-0.5.0.drv-0/grive2-ae06eccb38b2fe250c9ddeac3e3973f80b8a0aa9-src/libgrive/src/util/OS.cc(57): Throw in function void gr::os::Stat(const string&, gr::DateTime*, off_t*, bool*)
Dynamic exception type: boost::exception_detail::clone_impl<gr::os::Error>
[gr::expt::BacktraceTag*] = #0 0x4a21a9 :0 gr::Exception::Exception()
#1 0x49dda1 :0 gr::os::Error::Error()
#2 0x49dc39 :0 gr::os::Stat(boost::filesystem::path const&, gr::DateTime*, long*, bool*)
#3 0x468fcc :0 gr::Resource::FromLocal(gr::Val&)
#4 0x449ead :0 gr::State::FromLocal(boost::filesystem::path const&, gr::Resource*, gr::Val&)
#5 0x449eff :0 gr::State::FromLocal(boost::filesystem::path const&, gr::Resource*, gr::Val&)
#6 0x44a602 :0 gr::State::FromLocal(boost::filesystem::path const&)
#7 0x44782c :0 gr::Drive::DetectChanges()
#8 0x43a1cb :0 Main(int, char**)
#9 0x4385fb :0 main
#10 0x7ff2e6cc1985 /nix/store/pv9sza1cf2kpawck7wbwdnhlip5h57lg-glibc-2.23/lib64/libc.so.6 __libc_start_main
#11 0x4391e9 :0 _start

2, "No such file or directory"[boost::errinfo_file_name_*] = ./CellDynamics(J. Simoes)/.#clustering-article.org
[boost::errinfo_api_function_*] = stat
```

joachifm commented:
Thanks, that is quite useful. Adding binutils is clearly the right thing to do, but I'd like to get to the bottom of why stdenv is needed, so I'll leave this open for now.

joachifm commented on pkgs/tools/filesystems/grive2/default.nix:
```diff
@@ -0,0 +1,25 @@
+    sha256 = "0gyic9228j25l5x8qj9xxxp2cgbw6y4skxqx0xrq6qilhv4lj23c";
+  };
+
+  nativeBuildInputs = [cmake pkgconfig];
```
The usual style is to leave a space after/before brackets, as in `[ a b c d e ]`.

joachifm commented on pkgs/tools/filesystems/grive2/default.nix:
```diff
@@ -0,0 +1,25 @@
+  buildInputs = [libgcrypt yajl curl expat stdenv boost binutils];
+
+  meta = {
+    description = "An open source Linux client for Google Drive";
```
I think simply "Google Drive client" suffices as a description; the platform and license information is already provided by their respective attributes.

joachifm commented on pkgs/tools/filesystems/grive2/default.nix:
```diff
@@ -0,0 +1,25 @@
+
+  meta = {
+    description = "An open source Linux client for Google Drive";
+    homepage =https://github.com/vitalif/grive2;
```
For symmetry, leave a space after the equals sign here

joachifm commented on pkgs/tools/filesystems/grive2/default.nix:
```diff
@@ -0,0 +1,25 @@
+  src = fetchFromGitHub {
+    owner = "vitalif";
+    repo = "grive2";
+    rev =  "ae06eccb38b2fe250c9ddeac3e3973f80b8a0aa9";
```
You can do "v${version}" as the revision here (fetchFromGitHub takes the information provided and turns it into a URL to a tarball generated by github).

joachifm commented on pkgs/tools/filesystems/grive2/default.nix:
```diff
@@ -0,0 +1,25 @@
+
+  buildInputs = [libgcrypt yajl curl expat stdenv boost binutils];
+
+  meta = {
```
A common idiom is to use `stdenv.lib` here to avoid having to type out `stdenv.lib.licenses` etc.

joachifm commented:
Added a few more nitpicks as well

guaraqe commented:
Ok, I followed your suggestions, thanks for the tips!

I'm not a C++ programmer, but could be the problem be due to version incompability between the overlapping functionalities of stdenv and binutils?

joachifm commented:
I'd think the version of binutils in stdenv and the one you're passing in would be the same. The stack trace does not indicate a run-time linker error, looks to me more like an uncaught exception.

To elaborate (I apologize if you already know this): The purpose of stdenv is to provide an environment for the builder script. Adding something to `buildInputs` generally means that any binaries are added to `PATH`, any libraries are added to the linker path, and so on, so that the compiler and other tools necessary to perform the build can make use of it. This is why adding stdenv to the inputs ought to be redundant, it shouldn't add anything not already present in the builder environment.

To elaborate even more: adding someting to `buildInputs` is not a mechanism for directly controlling the run-time dependencies of an executable (or build production, more generally). Because all store paths are unique, nix actually scans the output productions for store paths and uses that information to compute a precise listing of run-time dependencies (aka the closure). Inputs only end up in the closure if the build process ends up writing their paths into the productions (or you write the path somewhere in the output manually). Nix knows what dynamic libraries an executable needs because their paths are written into the executable at build-time. Now, a build script may falsely ignore a missing run-time dependency (i.e., hasn't been added to inputs and so is unavailable in the build environment), creating an executable that doesn't have all its required libraries in its run-path, resulting in a crash when the symbol(s) in question are resolved during execution, but that doesn't appear to be the case here.

jagajaga commented:
Ready to merge?

guaraqe commented:
The tests worked, but @joachifm is the judge for the merging.

jagajaga commented:
Please squash commits.

guaraqe commented:
Is this ok?
